lib/Core/Executor.cpp

```diff
@@ -4060,53 +4047,34 @@ ExecutionState *Executor::executeCheckPersistence(ExecutionState &state,
   const PersistentState *ps = dyn_cast<PersistentState>(os);
   assert(ps);
 
-  ExecutionState *currState = &state;
-
-  for (const ref<Expr> &sliceConstraint : ps->getConstraints()) {
-    if (!currState) break;
+  // Get a symbolic offset into the object and constrain it to be within
+  // the object's bounds.
+  auto anyOffset = ps->getAnyOffsetExpr();
+  auto inBoundsConstraint = mo->getBoundsCheckOffset(anyOffset);
+  state.constraints.addConstraint(inBoundsConstraint);
 
-    // fprintf(stderr, "check slice for %p\n", (void*)mo->address);
-    // sliceConstraint->dump();
-    currState->constraints.addConstraint(sliceConstraint);
+  StatePair isPersisted = fork(state, ps->getIsOffsetPersistedExpr(anyOffset), true);
 
-    StatePair isPersisted = fork(*currState, ps->isPersistedUnconstrained(), true);
+  if (isPersisted.first) {
+    isPersisted.first->constraints.removeConstraint(inBoundsConstraint);
+  }
+  if (isPersisted.second) {
+    isPersisted.second->constraints.removeConstraint(inBoundsConstraint);
+  } 
 
-    if (isPersisted.first) {
-      isPersisted.first->constraints.removeConstraint(sliceConstraint);
-    }
-    if (isPersisted.second) {
-      isPersisted.second->constraints.removeConstraint(sliceConstraint);
-    } 
-
-    // If there's a state where it's not definitely persisted, terminate it.
-    ExecutionState *notPersisted = isPersisted.second;
-    if (notPersisted) {
-      const ObjectState *nos = notPersisted->addressSpace.findObject(mo);
-      assert(nos);
-      const PersistentState *nps = dyn_cast<PersistentState>(nos);
-      assert(nps);
-
-      // for (const ref<Expr> cons : notPersisted->constraints) {
-      //   cons->dump();
-      // }
-      // klee_warning("Non-persistence detected!");
-      // Should instead do the root cause analysis.
-      // std::string' addrInfo("\npmem persistence failures:\n");
-      // mo->getAllocInfo(addrInfo);
-      // uint64_t id = 1;
-      // for (const auto &str : ) {
-      //   addrInfo += std::to_string(id++) + ") " + str + std::string("\n");
-      // }
-      // klee_warning("addrInfo: '%s'", addrInfo.c_str());
-
-      terminateStateOnPmemError(*notPersisted, 
-                                nps->getRootCauses(solver, *notPersisted));
-    }
+  // If there's a state where it's not definitely persisted, terminate it.
+  ExecutionState *notPersisted = isPersisted.second;
+  if (notPersisted) {
+    const ObjectState *nos = notPersisted->addressSpace.findObject(mo);
+    assert(nos);
+    const PersistentState *nps = dyn_cast<PersistentState>(nos);
+    assert(nps);
 
-    currState = isPersisted.first;
-  } 
+    terminateStateOnPmemError(*notPersisted, 
+                              nps->getRootCauses(solver, *notPersisted));
+  }
 
-  return currState;
+  return &state;
```

Comment:
Is `state == *isPersisted.first` guaranteed to always be true? I remember something saying that `fork` returns `&state` as one of the `ExecutionState`s in the `StatePair`, but I don't know of the other specifics.

Reply:
It wasn't clear to me what the return value of this function was supposed to be. Can you clarify?

Reply:
I think it should be the good state.